lib/schmooze/base.rb

```diff
@@ -41,85 +43,118 @@ def finalize(stdin, stdout, stderr, process_thread)
     end
 
     def initialize(root, env={})
-      @_schmooze_env = env
-      @_schmooze_root = root
-      @_schmooze_code = ProcessorGenerator.generate(self.class.instance_variable_get(:@_schmooze_imports) || [], self.class.instance_variable_get(:@_schmooze_methods) || [])
+      @_schmooze_bridge = Bridge.new(
+        env: env,
+        root: root,
+        schmooze_class: self.class
+      )
     end
 
     def pid
-      @_schmooze_process_thread && @_schmooze_process_thread.pid
+      @_schmooze_bridge.process_thread_pid
     end
 
     private
       def ensure_process_is_spawned
-        return if @_schmooze_process_thread
-        spawn_process
+        return if @_schmooze_bridge.process_thread
+        @_schmooze_bridge.spawn_process
+      end
+
+      def call_js_method(method, args)
+        ensure_process_is_spawned
+
+        @_schmooze_bridge.stdin.puts JSON.dump([method, args])
+        input = @_schmooze_bridge.stdout.gets
+        raise Errno::EPIPE, "Can't read from stdout" if input.nil?
+
+        status, message, error_class = JSON.parse(input)
+
+        if status == 'ok'
+          message
+        else
+          if error_class.nil?
+            raise Schmooze::JavaScript::UnknownError, message
+          else
+            raise Schmooze::JavaScript.const_get(error_class, false), message
+          end
+        end
+      rescue Errno::EPIPE, IOError
+        # TODO(bouk): restart or something? If this happens the process is completely broken
+        raise ::StandardError, "Schmooze process failed:\n#{@_schmooze_bridge.stderr.read}"
+      end
+
+    class Declarations
+      attr_accessor :imports_list, :methods_list
+
+      def initialize(imports_list: [], methods_list: [])
+        @imports_list = imports_list
+        @methods_list = methods_list
+      end
+    end
+
+    class Bridge
+      attr_accessor :env, :root, :code, :process_thread, :stdin, :stdout, :stderr
+
+      def initialize(env: nil, root: nil, schmooze_class: nil)
```

Comment:
I'm curious: what versions of Ruby did you want this gem to support? If it's 2.1+, I can remove these default arguments since these are all required. If it's < 2.0 I can change this to an options hash.